src/sentry/preprod/eap/read.py

```diff
@@ -0,0 +1,300 @@
+
+from google.protobuf.timestamp_pb2 import Timestamp
+from sentry_protos.snuba.v1.downsampled_storage_pb2 import DownsampledStorageConfig
+from sentry_protos.snuba.v1.endpoint_trace_item_table_pb2 import (
+    Column,
+    TraceItemTableRequest,
+    TraceItemTableResponse,
+)
+from sentry_protos.snuba.v1.request_common_pb2 import (
+    PageToken,
+    RequestMeta,
+    TraceItemType,
+)
+from sentry_protos.snuba.v1.trace_item_attribute_pb2 import AttributeKey, AttributeValue
+from sentry_protos.snuba.v1.trace_item_filter_pb2 import (
+    AndFilter,
+    ComparisonFilter,
+    TraceItemFilter,
+)
+
+from sentry.utils import snuba_rpc
+
+
+@dataclass
+class PreprodSizeFilters:
+    # IDs
+    artifact_id: int | None = None
+
+    # App/Artifact attributes
+    app_id: str | None = None
+    artifact_type: int | None = None
+
+    # Build configuration
+    build_configuration_name: str | None = None
+
+    # Git attributes
+    git_head_ref: str | None = None
+    git_head_sha: str | None = None
+    git_base_ref: str | None = None
+    git_base_sha: str | None = None
+    git_provider: str | None = None
+    git_head_repo_name: str | None = None
+    git_base_repo_name: str | None = None
+    git_pr_number: int | None = None
+
+
+_FIELD_TYPES: dict[str, AttributeKey.Type.ValueType] = {
+    "app_id": AttributeKey.Type.TYPE_STRING,
+    "artifact_type": AttributeKey.Type.TYPE_INT,
+    "build_configuration_name": AttributeKey.Type.TYPE_STRING,
+    "git_head_ref": AttributeKey.Type.TYPE_STRING,
+    "git_head_sha": AttributeKey.Type.TYPE_STRING,
+    "git_base_ref": AttributeKey.Type.TYPE_STRING,
+    "git_base_sha": AttributeKey.Type.TYPE_STRING,
+    "git_provider": AttributeKey.Type.TYPE_STRING,
+    "git_head_repo_name": AttributeKey.Type.TYPE_STRING,
+    "git_base_repo_name": AttributeKey.Type.TYPE_STRING,
+    "git_pr_number": AttributeKey.Type.TYPE_INT,
+    # Note: artifact_id is handled specially (converted to trace_id)
+}
+
+
+def query_preprod_size_metrics(
+    organization_id: int,
+    project_ids: list[int],
+    start: datetime,
+    end: datetime,
+    filters: PreprodSizeFilters | None = None,
+    limit: int = 100,
+    offset: int = 0,
+) -> TraceItemTableResponse:
+    """
+    Query preprod size metrics from EAP.
+
+    Example:
+        # Query all iOS Release builds on main branch for a specific app
+        query_preprod_size_metrics(
+            organization_id=1,
+            project_ids=[1, 2],
+            start=datetime.now() - timedelta(days=30),
+            end=datetime.now(),
+            filters=PreprodSizeFilters(
+                app_id="com.example.app",
+                artifact_type=0,  # iOS
+                build_configuration_name="Release",
+                git_head_ref="main",
+            ),
+        )
+    """
+    start_timestamp = Timestamp()
+    start_timestamp.FromDatetime(start)
+
+    end_timestamp = Timestamp()
+    end_timestamp.FromDatetime(end)
+
+    columns = [
```

Comment:
For now it's requesting all columns but we can make this configurable as a param too.